File: src/leetcode/java/medium/ConstructBinaryTreeFromInorderAndPostorderTraversal_106.java
package leetcode.java.medium;

import java.util.Stack;

public class ConstructBinaryTreeFromInorderAndPostorderTraversal_106 {
  public TreeNode buildTree(int[] inorder, int[] postorder) {
    if (inorder.length == 0 || postorder.length == 0 || inorder.length != postorder.length) {
      return null;
    }
    int postPt = postorder.length - 1;
    int inPt = inorder.length - 1;
    TreeNode root = new TreeNode(postorder[postPt]);
    Stack<TreeNode> stack = new Stack<>();
    stack.push(root);
    postPt--;

    TreeNode node;
    while (true) {
      if (inorder[inPt] == stack.peek().val) {
        if (--inPt < 0) break;
        node = stack.pop();
        if (!stack.isEmpty() && stack.peek().val == inorder[inPt]) {
          continue;
        }
        node.left = new TreeNode(postorder[postPt]);
        stack.push(node.left);
      } else {
        node = new TreeNode(postorder[postPt]);
        stack.peek().right = node;
        stack.push(node);
      }
      postPt--;
    }
    return root;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}

File: src/leetcode/python/ConstructBinaryTreeFromInorderAndPostorderTraversal_106.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class ConstructBinaryTreeFromInorderAndPostorderTraversal_106(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if not inorder or not postorder:
            return None

        root = TreeNode(postorder.pop())
        val_idx = inorder.index(root.val)

        root.right = self.buildTree(inorder[val_idx+1:], postorder)
        root.left = self.buildTree(inorder[:val_idx], postorder)
        return root

